api/types/common/common.go

```diff
@@ -0,0 +1,434 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"regexp"
+	"time"
+
+	"github.com/gravitational/trace"
+	"golang.org/x/exp/maps"
+	"golang.org/x/exp/slices"
+
+	commonpb "github.com/gravitational/teleport/api/gen/proto/go/common/v1"
+	"github.com/gravitational/teleport/api/utils"
+)
+
+// Builder is an interface that can build a resource object.
+type Builder[T any] interface {
```

Comment:
I'm a fan of keeping public API surface area as small and easy to comprehend as possible.

Since the API module is intended for outside use, once you add something, you're stuck supporting for a long time.

Can we do without this new builder pattern and new set of exported types?

Reply:
We could definitely get rid of it -- it's not essential -- but I'd like to argue for it for a bit with an example:

https://github.com/gravitational/teleport/pull/27812

So, I had to modify the `NewUserGroup`  function here to support this feature, as before the user group was basically just a name and nothing else. As a result, it broke the `e` compilation. It wasn't a huge deal to fix. If I wanted to avoid the `e` breakage, I could have done so by introducing an extra `NewUserGroup` function that took the spec and then doing a bit of a dance to try to get rid of that function, or just keep the extra function around.

However, a builder pattern would have avoided that breakage entirely and it would have eliminated the need to submit multiple PRs to try to clean up (if I were so inclined). Admittedly the reasoning for the builder pattern is to make our lives easier w/r/t `e` changes, but I think it's worth considering.

That being said, if you're opposed to it I can remove it.

Reply:
On second thought I suppose this wouldn't be an issue if we just didn't have a `NewAccessList` function altogether, either.

Reply:
After some discussion on https://github.com/gravitational/teleport/pull/28386, I'll remove the builders.

Reply:
After some discussion on https://github.com/gravitational/teleport/pull/28386, I'll remove the builders.